packages/flutter/lib/src/widgets/two_dimensional_scroll_view.dart

```diff
@@ -0,0 +1,194 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/gestures.dart';
+import 'package:flutter/rendering.dart';
+
+import 'focus_manager.dart';
+import 'focus_scope.dart';
+import 'framework.dart';
+import 'notification_listener.dart';
+import 'primary_scroll_controller.dart';
+import 'scroll_controller.dart';
+import 'scroll_delegate.dart';
+import 'scroll_notification.dart';
+import 'scroll_physics.dart';
+import 'scroll_view.dart';
+import 'scrollable.dart';
+import 'scrollable_helpers.dart';
+import 'two_dimensional_viewport.dart';
+
+/// A widget that combines a [TwoDimensionalScrollable] and a
+/// [TwoDimensionalViewport] to create an interactive scrolling pane of content
+/// in both vertical and horizontal dimensions.
+///
+/// A two-way scrollable widget consist of three pieces:
+///
+///  1. A [TwoDimensionalScrollable] widget, which listens for various user
+///     gestures and implements the interaction design for scrolling.
+///  2. A [TwoDimensionalViewport] widget, which implements the visual design
+///     for scrolling by displaying only a portion
+///     of the widgets inside the scroll view.
+///  3. A [TwoDimensionalChildDelegate], which provides the children visible in
+///     the scroll view.
+///
+/// [TwoDimensionalScrollView] helps orchestrate these pieces by creating the
+/// [TwoDimensionalScrollable] and deferring to its subclass to implement
+/// [viewportBuilder], which builds a subclass of [TwoDimensionalViewport]. The
+/// [TwoDimensionalChildDelegate] is provided by the [delegate] parameter.
+///
+/// A [TwoDimensionalScrollView] has two different [ScrollPosition]s, one for
+/// each [Axis]. This means that there are also two unique [ScrollController]s
+/// for these positions. To provide a ScrollController to access the
+/// ScrollPosition, use the [ScrollableDetails.controller] property of the
+/// associated axis that is provided to this scroll view.
+abstract class TwoDimensionalScrollView extends StatelessWidget {
```

Comment:
@goderbauer WDYT of this being concrete with a `viewportBuilder` parameter? That would mean folks only need to focus on implementing the 2D viewport render object subclass.

Reply:
That sounds reasonable to me.

Reply:
Actually, this was not feasible.

Reply:
Many of the parameters of 2D scroll view need to be passed to the 2D viewport, so a `viewportBuilder` would either have an ugly signature, or the code for the user would be pretty ugly.